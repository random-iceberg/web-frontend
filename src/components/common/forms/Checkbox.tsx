import React from "react";

/**
 * Props for the Checkbox component.
 * @typedef {object} CheckboxProps
 * @property {string} label - Text label displayed next to the checkbox.
 * @property {string} id - Unique ID for the checkbox input and linking label.
 * @property {boolean} checked - Whether the checkbox is currently checked.
 * @property {(checked: boolean) => void} onChange - Callback function when the checkbox state changes.
 * @property {boolean} [disabled=false] - Whether the checkbox is disabled.
 * @property {string} [description] - Optional description text displayed below the label.
 */
type CheckboxProps = {
  label: string;
  id: string;
  checked: boolean;
  onChange: (checked: boolean) => void;
  disabled?: boolean;
  description?: string; // Optional description text
};

/**
 * A styled checkbox input component with a label and optional description.
 *
 * @component
 * @param {CheckboxProps} props - Props for the Checkbox component.
 * @example
 * const [isChecked, setIsChecked] = useState(false);
 *
 * <Checkbox
 *   id="agree"
 *   label="I agree to the terms"
 *   checked={isChecked}
 *   onChange={setIsChecked}
 *   description="Please read the terms carefully."
 * />
 */
const Checkbox: React.FC<CheckboxProps> = ({
  label,
  id,
  checked,
  onChange,
  disabled = false,
  description,
}) => {
  const descriptionId = description ? `${id}-description` : undefined;
  return (
    <div className="flex items-start">
      {" "}
      {/* Changed to items-start for better alignment with description */}
      <div className="flex h-5 items-center">
        <input
          type="checkbox"
          id={id}
          checked={checked}
          onChange={(e) => onChange(e.target.checked)}
          disabled={disabled}
          className="h-4 w-4 rounded border-gray-300 text-primary focus:ring-primary disabled:opacity-50 disabled:cursor-not-allowed"
          aria-describedby={descriptionId}
        />
      </div>
      <div className="ml-3 text-base">
        <label
          htmlFor={id}
          className={`font-medium ${disabled ? "text-gray-500" : "text-gray-700"}`}
        >
          {label}
        </label>
        {description && (
          <p id={descriptionId} className="mt-1 text-sm text-gray-600">
            {description}
          </p>
        )}
      </div>
    </div>
  );
};

export default Checkbox; // Exporting as Checkbox
