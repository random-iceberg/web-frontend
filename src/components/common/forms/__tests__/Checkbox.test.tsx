import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Checkbox from "../Checkbox";

describe("Checkbox Component", () => {
  const mockOnChange = jest.fn();

  beforeEach(() => {
    mockOnChange.mockClear();
  });

  it("renders correctly when unchecked", () => {
    const { asFragment } = render(
      <Checkbox
        id="test-checkbox"
        label="Test Checkbox"
        checked={false}
        onChange={mockOnChange}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when checked", () => {
    const { asFragment } = render(
      <Checkbox
        id="test-checkbox"
        label="Test Checkbox"
        checked={true}
        onChange={mockOnChange}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when disabled and unchecked", () => {
    const { asFragment } = render(
      <Checkbox
        id="test-checkbox"
        label="Test Checkbox"
        checked={false}
        onChange={mockOnChange}
        disabled
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when disabled and checked", () => {
    const { asFragment } = render(
      <Checkbox
        id="test-checkbox"
        label="Test Checkbox"
        checked={true}
        onChange={mockOnChange}
        disabled
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with a description", () => {
    const { asFragment } = render(
      <Checkbox
        id="test-checkbox"
        label="Test Checkbox"
        checked={false}
        onChange={mockOnChange}
        description="This is a test description."
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onChange handler when clicked", () => {
    render(
      <Checkbox
        id="test-checkbox"
        label="Test Checkbox"
        checked={false}
        onChange={mockOnChange}
      />,
    );
    fireEvent.click(screen.getByLabelText("Test Checkbox"));
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(true);
  });

  it("calls onChange handler with correct value when already checked", () => {
    render(
      <Checkbox
        id="test-checkbox"
        label="Test Checkbox"
        checked={true}
        onChange={mockOnChange}
      />,
    );
    fireEvent.click(screen.getByLabelText("Test Checkbox"));
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(false);
  });

  it("does not call onChange handler when disabled", () => {
    render(
      <Checkbox
        id="test-checkbox"
        label="Test Checkbox"
        checked={false}
        onChange={mockOnChange}
        disabled
      />,
    );
    fireEvent.click(screen.getByLabelText("Test Checkbox"));
    expect(mockOnChange).not.toHaveBeenCalled();
  });
});
