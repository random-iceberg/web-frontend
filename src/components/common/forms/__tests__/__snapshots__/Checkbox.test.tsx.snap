// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component renders correctly when checked 1`] = `
<DocumentFragment>
  <div
    class="flex items-start"
  >
     
    <div
      class="flex h-5 items-center"
    >
      <input
        checked=""
        class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-primary disabled:opacity-50 disabled:cursor-not-allowed"
        id="test-checkbox"
        type="checkbox"
      />
    </div>
    <div
      class="ml-3 text-base"
    >
      <label
        class="font-medium text-gray-700"
        for="test-checkbox"
      >
        Test Checkbox
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Checkbox Component renders correctly when disabled and checked 1`] = `
<DocumentFragment>
  <div
    class="flex items-start"
  >
     
    <div
      class="flex h-5 items-center"
    >
      <input
        checked=""
        class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-primary disabled:opacity-50 disabled:cursor-not-allowed"
        disabled=""
        id="test-checkbox"
        type="checkbox"
      />
    </div>
    <div
      class="ml-3 text-base"
    >
      <label
        class="font-medium text-gray-500"
        for="test-checkbox"
      >
        Test Checkbox
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Checkbox Component renders correctly when disabled and unchecked 1`] = `
<DocumentFragment>
  <div
    class="flex items-start"
  >
     
    <div
      class="flex h-5 items-center"
    >
      <input
        class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-primary disabled:opacity-50 disabled:cursor-not-allowed"
        disabled=""
        id="test-checkbox"
        type="checkbox"
      />
    </div>
    <div
      class="ml-3 text-base"
    >
      <label
        class="font-medium text-gray-500"
        for="test-checkbox"
      >
        Test Checkbox
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Checkbox Component renders correctly when unchecked 1`] = `
<DocumentFragment>
  <div
    class="flex items-start"
  >
     
    <div
      class="flex h-5 items-center"
    >
      <input
        class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-primary disabled:opacity-50 disabled:cursor-not-allowed"
        id="test-checkbox"
        type="checkbox"
      />
    </div>
    <div
      class="ml-3 text-base"
    >
      <label
        class="font-medium text-gray-700"
        for="test-checkbox"
      >
        Test Checkbox
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Checkbox Component renders with a description 1`] = `
<DocumentFragment>
  <div
    class="flex items-start"
  >
     
    <div
      class="flex h-5 items-center"
    >
      <input
        aria-describedby="test-checkbox-description"
        class="h-4 w-4 rounded border-gray-300 text-primary focus:ring-primary disabled:opacity-50 disabled:cursor-not-allowed"
        id="test-checkbox"
        type="checkbox"
      />
    </div>
    <div
      class="ml-3 text-base"
    >
      <label
        class="font-medium text-gray-700"
        for="test-checkbox"
      >
        Test Checkbox
      </label>
      <p
        class="mt-1 text-sm text-gray-600"
        id="test-checkbox-description"
      >
        This is a test description.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
